File: challenges/CC_03_Snake_game_p5.js/snake.js
// Daniel Shiffman
// http://codingrainbow.com
// http://patreon.com/codingrainbow
// Code for: https://youtu.be/AaGK-fj-BAM

function Snake() {
  this.x = 0;
  this.y = 0;
  this.xspeed = 1;
  this.yspeed = 0;
  this.total = 0;
  this.tail = [];
  this.flag = false;
  
  this.lvl;
  this.lvl_limit = 10 ;
  
  this.currentlvl;
  this.turboflag;
  
  this.eat = function(pos) {
    var d = dist(this.x, this.y, pos.x, pos.y);
    if (d < 1) {
      this.total++;
	  this.flag = true;
      return true;
    } else {
      return false;
    }
  }

  this.dir = function(x, y) {
    this.xspeed = x;
    this.yspeed = y;
  }

  this.Reset = function() {
		// reset score 
        this.total = 0;
		this.lvl = num_lvl;
		this.currentlvl= num_lvl;
		// reser snake
        this.tail = [];
		this.flag = false;
		// reset position
		this.x = 0;
		this.y = 0;
		this.xspeed = 1;
		this.yspeed = 0;
		pickLocation();			
		//reset turbo
		turboflag = true;
  }
  
  this.ResetSnake = function() {
		alert("Game over");
		this.Reset();	
	}

  this.EatTail = function(){
	for (var i = 0; i < this.tail.length; i++) {
      var pos = this.tail[i];
      var d = dist(this.x, this.y, pos.x, pos.y);
      if (d < 1) {
		 this.ResetSnake();
		 return true;
      } 
    }
	return false ;
  }
  
  this.HitWalls = function(walls){
	for (var i = 0; i < walls.length; i++) {
      var d = dist(this.x, this.y, walls[i].x, walls[i].y);
      if (d < 1) {
		 this.ResetSnake();
		 return true;
      } 
    }
	return false ;
  }
  
  this.death = function(walls) {
	var dead1 = false;
	var dead2 = false;
	
	dead1 = this.EatTail();
	dead2 = this.HitWalls(walls);
	
	return (dead2 | dead1) ;
  }

  this.update = function() {
    if (this.total === this.tail.length) {
      for (var i = 0; i < this.tail.length - 1; i++) {
        this.tail[i] = this.tail[i + 1];
      }
    }
    this.tail[this.total - 1] = createVector(this.x, this.y);

    this.x = this.x + this.xspeed * scl;
    this.y = this.y + this.yspeed * scl;

	// cross the wall
	if ( this.x > width- scl) {
		this.x = 0;	
		
	}
	else if ( this.x < 0) {
		this.x =  width- scl;	
	}
	
	if ( this.y > lower_edge- scl) {
		this.y = 0;	
	}
	else if ( this.y < 0) {
		this.y =  lower_edge- scl;	
	}

  }
  
  this.turbo = function() {
	  if(turboflag)this.currentlvl = this.lvl;
	  turboflag = false;
	  this.lvl = ceil(this.lvl/2);
  }
  
  this.normal = function() {
	  this.lvl = this.currentlvl;
	  turboflag = true;
  }
  
  this.show = function() {
    fill(255);
    for (var i = 0; i < this.tail.length; i++) {
      rect(this.tail[i].x, this.tail[i].y, scl, scl);
    }
    rect(this.x, this.y, scl, scl);

  }
}

File: CodingChallenges/CC_03_Snake_game_py/Snake.py
# Daniel Shiffman - Translated  to Python by Ali Al Saqban
# http://codingtra.in
# http://patreon.com/codingtrain

# The Snake Game
# https://youtu.be/AaGK-fj-BAM

class Snake(object):
    
    def __init__(self,scl):
        self.x = 0
        self.y = 0
        self.xspeed = 1
        self.yspeed = 0
        self.total = 0
        self.tail = []
        self.scl = scl
        
    def eat(self, pos):
        d = dist(self.x, self.y, pos.x, pos.y)
        if(d < 1):
            self.total += 1
            return True
        else:
            return False
    
    def dir(self, x, y):
        self.xspeed = x
        self.yspeed = y
    
    def death(self):
        for i in self.tail:
            pos = i
            d = dist(self.x,self.y,pos.x,pos.y)
            if(d < 1):
                print("starting over")
                self.total = 0
                self.tail = []
    
    def update(self):
        print(str(self.total) + " " + str(len(self.tail)))
        if(self.total > 0):
            if(self.total == len(self.tail) and not len(self.tail) == 0):
                self.tail.pop(0)
            self.tail.append(PVector(self.x,self.y))
        
        self.x = self.x + self.xspeed*self.scl
        self.y = self.y + self.yspeed*self.scl
        
        self.x = constrain(self.x, 0, width-self.scl)
        self.y = constrain(self.y, 0, height-self.scl)

    def show(self):
        fill(255)
        for v in self.tail:
            rect(v.x, v.y, self.scl, self.scl)
        
        rect(self.x, self.y, self.scl, self.scl)